src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,456 @@
+    return an instance of their subclass from calls into methods.
+    """
+
+    def __init__(self):
+        pass
+
+    @property
+    def is_persistent(self):
+        """
+        Has the operation updated the orchestrator's configuration
+        persistently?  Typically this would indicate that an update
+        had been written to a manifest, but that the update
+        had not necessarily been pushed out to the cluster.
+        """
+        raise NotImplementedError()
+
+    @property
+    def is_effective(self):
+        """
+        Has the operation taken effect on the cluster?  For example,
+        if we were adding a service, has it come up and appeared
+        in Ceph's cluster maps?
+        """
+        raise NotImplementedError()
+
+    @property
+    def is_complete(self):
+        return self.is_errored or (self.is_persistent and self.is_effective)
+
+    @property
+    def is_read(self):
+        return False
+
+
+class Orchestrator(object):
+    """
+    Calls in this class may do long running remote operations, with time
+    periods ranging from network latencies to package install latencies and large
+    internet downloads.  For that reason, all are asynchronous, and return
+    ``Completion`` objects.
+
+    Implementations are not required to start work on an operation until
+    the caller waits on the relevant Completion objects.  Callers making
+    multiple updates should not wait on Completions until they're done
+    sending operations: this enables implementations to batch up a series
+    of updates when wait() is called on a set of Completion objects.
+
+    Implementations are encouraged to keep reasonably fresh caches of
+    the status of the system: it is better to serve a stale-but-recent
+    result read of e.g. device inventory than it is to keep the caller waiting
+    while you scan hosts every time.
+    """
+
+    def is_orchestrator_module(self):
+        """
+        Enable other modules to interrogate this module to discover
+        whether it's usable as an orchestrator module.
+
+        Subclasses do not need to override this.
+        """
+        return True
+
+    def available(self):
+        """
+        Report whether we can talk to the orchestrator.  This is the
+        place to give the user a meaningful message if the orchestrator
+        isn't running or can't be contacted.
+
+        This method may be called frequently (e.g. every page load
+        to conditionally display a warning banner), so make sure it's
+        not too expensive.  It's okay to give a slightly stale status
+        (e.g. based on a periodic background ping of the orchestrator)
+        if that's necessary to make this method fast.
+
+        Do not override this method if you don't have a meaningful
+        status to return: the default None, None return value is used
+        to indicate that a module is unable to indicate its availability.
+
+        @return two-tuple of boolean, string
```

Comment:
this `@return` is not a valid RST directive for docstrings, maybe you wanted `:return:` ?

Reply:
Some tools & IDEs understand something like
```
:rtype: (bool, str) 
```
and can generate type-hints or warnings.